diff --git a/examples/mnist/train.py b/examples/mnist/train.py
index e93321f..6c77bf3 100644
--- a/examples/mnist/train.py
+++ b/examples/mnist/train.py
@@ -85,8 +85,8 @@ def create_model(key):
   return model
 
 
-def create_optimizer(model, learning_rate, beta):
-  optimizer_def = optim.Momentum(learning_rate=learning_rate, beta=beta)
+def create_fast_optimizer(model, learning_rate, beta):
+  optimizer_def = optim.FastMomentum(learning_rate=learning_rate, beta=beta)
   optimizer = optimizer_def.create(model)
   return optimizer
 
