diff --git a/examples/mnist/train.py b/examples/mnist/train.py
index 5dc74db..73aeac8 100644
--- a/examples/mnist/train.py
+++ b/examples/mnist/train.py
@@ -134,7 +134,7 @@ def train_epoch(optimizer, train_ds, batch_size, epoch, rng):
   """Train for a single epoch."""
   train_ds_size = len(train_ds['image'])
   steps_per_epoch = train_ds_size // batch_size
-
+  
   perms = rng.permutation(len(train_ds['image']))
   perms = perms[:steps_per_epoch * batch_size]  # skip incomplete batch
   perms = perms.reshape((steps_per_epoch, batch_size))
